### ISSUE TITLE
A Couple Of Issues

### ISSUE THREAD

**Issue by MOAMaster** [mikeoliphant/AudioPlugSharp#10]
I have two problems that I can't seem to resolve.
Maybe you can give me insight if you have seen them before.

1. Cakewalk suddenly started giving me a load error:
    AudioPlugSharVst, GetPluginFactory
![LoadError](https://user-images.githubusercontent.com/20228570/151088715-e328c58c-a677-457d-8157-e8463a392fd1.jpg)
   I have attached a screen dump. I tried cleaning/rebuilding, copying and
  renaming into project folders, nothing fixes the problem. 
  Any idea why I would get this problem and how to fix it?

2. The second problem is a bit more insidious.
   I have a bunch of simple apps in which I override GetEditorWindow.
   For all of them, I use WPF controls.
  So, I decided to do a custom user control.
        I found a really nice knob control on the web in xaml.
        Converted to list to a named canvas.
       so I created a User Control. I created knob programmatically
       in C#. Then made the knob a child of the named canvas.
       I put the code the generates the knob into a WPF DLL and
      that CODE calls the rotary control DLL to setup the handlers.
      Wrote a test driver. Works great, I can move the pointers around
      and read the dial values.

      Then, I integrated into a plugin.
      The knob displays beautifully.
      But, when I tried to read the value of the knob, I am told:
      can't access because that control is on a different thread from the
         calling thread.

     I have tried everything I can find to defeat this: using dispatcher,
     I get a null dispatcher, I finally seemed to find a way to get a non-null,
     dispatcher but the call to get the value returns nothing!

     I see you have done lots of plugins. Have you ever had this problem?
     How did you get around it? I looked through your code but don't see
     what you did. Any ideas?

     I have one other idea I will try -- putting the two classes into the 
      plugin   directly. That might fix the problem.  Even if that fixes, don't
      understand why separate DLLs are a problem.

     Any ideas would be much appreciated.

Thanks.

Bruce

**Comment by MOAMaster** [comment]
More on the threading issue:
In my test driver for the dial, I was able to replicate the cross threading problem. I was able to solve the problem by wrapping the access to the dial value:
                   this.Dispatcher.BeginInvoke(new Action(() =>
                     {
                         gain = GainDial.Value;
                     }));
This returned the value of the dial.
But, when I tried to use this in the plugin, the execution is successful,
but no value is returned!!
I have done some deep dive into the vstsdk and have determined that 
the plugin does seem to run in a thread. 
So, how are you getting past the cross thread lock in your custom controls.
Thanks.

Bruce

**Comment by MOAMaster** [comment]
OK, finally figured out how to make the new knob control thread safe.
1. Added a method in the knob control class to use Dispatcher.Invoke to wrap the access to the knob control properties. This is now owned by the knob control class which has thread affinity to the UI thread. But, the knob control can access its own value inside the knob control class, even if the access is from another thread.
2. Had to wrap the call to that method in a try/catch block. When I stopped the audio, the whole plugin hung. I guess the host keeps making calls to process which somehow hangs after stop audio.
3. Took me several days to figure out this simple issue. I tried every 
possible approach. I saw this in the example you sent me, so I tried.
